interactions.go

```diff
@@ -90,13 +96,86 @@ type Interaction struct {
 	Version int    `json:"version"`
 }
 
-// ApplicationCommandInteractionData contains data received in an interaction event.
+type interaction Interaction
+
+type rawInteraction struct {
+	interaction
+	Data json.RawMessage `json:"data"`
+}
+
+// UnmarshalJSON is a method for unmarshalling JSON object to Interaction.
+func (i *Interaction) UnmarshalJSON(raw []byte) (err error) {
+	var tmp rawInteraction
+	err = json.Unmarshal(raw, &tmp)
+	if err != nil {
+		return
+	}
+
+	*i = Interaction(tmp.interaction)
+
+	switch tmp.Type {
+	case InteractionApplicationCommand:
+		v := ApplicationCommandInteractionData{}
+		err = json.Unmarshal(tmp.Data, &v)
+		if err != nil {
+			return
+		}
+		i.Data = v
+	case InteractionMessageComponent:
+		v := MessageComponentInteractionData{}
+		err = json.Unmarshal(tmp.Data, &v)
+		if err != nil {
+			return
+		}
+		i.Data = v
+	}
+	return nil
+}
+
+// MessageComponentData is helper function to convert InteractionData to MessageComponentInteractionData.
+func (i Interaction) MessageComponentData() (data MessageComponentInteractionData) {
+	if i.Type != InteractionMessageComponent {
+		return
+	}
```

Comment:
I'm quite honestly not sure if this check is necessary. IMO, this pattern expects the developer to check that the interaction contains a component before calling MessageComponentData, so if they fail to do so, it's indicative of a developer error; this is what panics are for, and luckily the type assertion already panics in this situation. It could be worth a note in the comment about this behavior, but I think that this is the right way to go; returning the zero value is ripe for bugs.